File: packages/expo/ios/CreebaExpoModule.swift
import ExpoModulesCore
import Foundation

/**
 * Expo ↔ native bridge. All transport logic lives in `IrohNode` (iroh-ffi + mDNS
 * via Network.framework). This module only routes calls and events.
 *
 * App frames travel as JSON strings (identical to the desktop wire format): the
 * native side does not know the hello/data semantics.
 */
public class CreebaExpoModule: Module {
  private var node: IrohNode?

  public func definition() -> ModuleDefinition {
    Name("CreebaExpo")

    Events("onReady", "onPeerOpen", "onPeerClose", "onFrame", "onError")

    AsyncFunction("start") { (identityJson: String) -> String in
      let node = self.ensureNode()
      let publicKey = try await node.start(identityJson: identityJson)
      self.sendEvent("onReady", ["publicKey": publicKey])
      return publicKey
    }

    Function("setIdentity") { (identityJson: String) in
      let node = self.ensureNode()
      Task { await node.setIdentity(identityJson) }
    }

    Function("join") { (room: String) in
      let node = self.ensureNode()
      Task { await node.join(room: room) }
    }

    Function("send") { (peerId: String, frameJson: String) in
      guard let node = self.node else { return }
      Task { await node.send(peerId: peerId, frameJson: frameJson) }
    }

    Function("broadcast") { (frameJson: String) in
      guard let node = self.node else { return }
      Task { await node.broadcast(frameJson: frameJson) }
    }

    Function("destroy") {
      let node = self.node
      self.node = nil
      Task { await node?.destroy() }
    }

    OnDestroy {
      let node = self.node
      self.node = nil
      Task { await node?.destroy() }
    }
  }

  private func ensureNode() -> IrohNode {
    if let node = node { return node }
    let created = IrohNode { [weak self] name, body in
      // Native events must be dispatched on the module's thread.
      self?.sendEvent(name, body)
    }
    node = created
    return created
  }
}

File: packages/expo/android/src/main/java/com/creeba/expo/CreebaExpoModule.kt
package com.creeba.expo

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo ↔ native Android bridge. Transport logic lives in `IrohNode`
 * (iroh-ffi + NsdManager). App frames travel as JSON strings (wire format
 * identical to desktop).
 */
class CreebaExpoModule : Module() {
  private var node: IrohNode? = null

  override fun definition() = ModuleDefinition {
    Name("CreebaExpo")

    Events("onReady", "onPeerOpen", "onPeerClose", "onFrame", "onError")

    AsyncFunction("start") Coroutine { identityJson: String ->
      val n = ensureNode()
      val publicKey = n.start(identityJson)
      sendEvent("onReady", mapOf("publicKey" to publicKey))
      publicKey
    }

    Function("setIdentity") { identityJson: String ->
      ensureNode().setIdentity(identityJson)
    }

    Function("join") { room: String ->
      ensureNode().join(room)
    }

    Function("send") { peerId: String, frameJson: String ->
      node?.send(peerId, frameJson)
    }

    Function("broadcast") { frameJson: String ->
      node?.broadcast(frameJson)
    }

    Function("destroy") {
      node?.destroy()
      node = null
    }

    OnDestroy {
      node?.destroy()
      node = null
    }
  }

  private fun ensureNode(): IrohNode {
    return node ?: IrohNode(appContext.reactContext?.applicationContext) { name, body ->
      sendEvent(name, body)
    }.also { node = it }
  }
}
